### PR TITLE
3744: Rescue module for OpenList

### PR DESCRIPTION
**Pull request by martinb14**
#### Link to issue

https://platform.dandigbib.org/issues/3744

#### Description

A module for rescuing certain lists and elements

### CODE REVIEW

**Comment by cableman** [comment]
Please fill out the template given when creating new pull request (https://github.com/ding2/ding2/blob/master/.github/PULL_REQUEST_TEMPLATE.md). 

Please also look at the steps 3,4 and 5 in contributing file that where agreed upon when making this PR's to the project. See https://github.com/ding2/ding2/blob/master/.github/CONTRIBUTING.md

These guidelines exists to make the process of contributing to the project as fast and easy for all involved in the process and to give others an easy view of the changes for a given PR.

Please follow these guidelines.

**Comment by martinb14** [comment]
Link to issue
https://platform.dandigbib.org/issues/3744

Description
A module for rescuing certain lists and elements

Screenshot of the result
image


Checklist
YES: My complies with our coding guidelines.
YES: My code passes our static analysis suite. If not then I have added a comment explaining why this change should be exempt from the code standards and process.
YES: My code passes our continuous integration process. If not then I have added a comment explaining why this change should be exempt from the code standards and process.
image

Additional comments or questions
Im not sure I understand the checklist above. Isn't this what github does for us? Or does this imply that we have to run other test suites on our side?

**Comment by cableman** [comment]
If you use the template and [x] for each check we can see in the overview that you have locked through you code and check that it complies with the guidelines.

9/10 times both Scrutinizer and CI has not passed the checks for PR's and the core team then needs to go through these which is time consuming (when having 200+ PR's) and slows down the release process. May issues are return to the developer, just because the tests are not passed.

So used correctly we will get a much better overview (see the out-of 3/3 in the overview):

screen shot 2018-10-17 at 09 38 09

File: modules/p2/ding_list/modules/ding_listrescue/ding_listrescue.info [hunk 1]
```diff
@@ -0,0 +1,5 @@
+name = Ding list rescue
+description = Recreate missing lists and elements.
+core = 7.x
+package = Ding!
+dependencies[] = ding_list
```
Comment (cableman) [inline modules/p2/ding_list/modules/ding_listrescue/ding_listrescue.info]: Missing new line at the end of the file (see the red icon).

File: modules/p2/ding_list/modules/ding_listrescue/ding_listrescue.module [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+/**
+ * Rescue form submit.
+ */
+function ding_listrescue_form_submit($form, &$form_state) {
+  global $user;
+
+  $form_state['rebuild'] = TRUE;
+
+  // Load the user object.
+  $account = user_load($user->uid);
```
Comment (cableman) [inline modules/p2/ding_list/modules/ding_listrescue/ding_listrescue.module]: Why load the user account when you have the user object in $user?

File: modules/p2/ding_list/modules/ding_listrescue/ding_listrescue.module [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+  }
+  ding_listrescue_log("Done");
+
+  drupal_set_message(t("Your lists have been imported"));
+}
+
+/**
+ * Logging of actions
+ */
+function ding_listrescue_log($msg) {
```
Comment (cableman) [inline modules/p2/ding_list/modules/ding_listrescue/ding_listrescue.module]: Not sure why this logging with drupal set message is need?

File: modules/p2/ding_list/modules/ding_listrescue/ding_listrescue.module [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+  if (!empty($_GET['debug'])) {
+    drupal_set_message($msg);
+  }
+}
+
+/**
+ * Get local elements from a local list.
+ */
+function ding_listrescue_get_local_elements($local) {
+  $max_elements = 200;
```
Comment (cableman) [inline modules/p2/ding_list/modules/ding_listrescue/ding_listrescue.module]: Why the limit of 200?

File: modules/p2/ding_list/modules/ding_listrescue/ding_listrescue.module [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+  $element_query->addField('fw', 'field_weight_value', 'weight');
+
+  return $element_query->range(0, $max_elements)->execute()->fetchAll();
+}
+
+/**
+ * Get all the local lists.
+ */
+function ding_listrescue_get_local_lists($account = NULL, $types = NULL) {
+  $max_lists = 50;
```
Comment (cableman) [inline modules/p2/ding_list/modules/ding_listrescue/ding_listrescue.module]: Why the limit of 50?

File: modules/p2/ding_list/modules/ding_listrescue/ding_listrescue.module [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+
+  // Load account.
+  if (empty($account)) {
+    global $user;
+    $account = user_load($user->uid);    
+  }
```
Comment (cableman) [inline modules/p2/ding_list/modules/ding_listrescue/ding_listrescue.module]: Why the user load? The $user is the same as what you get from user_load?

File: modules/p2/ding_list/modules/ding_listrescue/ding_listrescue.module (hunk 3 above)
Comment (cableman) [reply modules/p2/ding_list/modules/ding_listrescue/ding_listrescue.module]: If required function. The function documentation is missing @param

File: modules/p2/ding_list/modules/ding_listrescue/ding_listrescue.module [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+ */
+function ding_listrescue_log($msg) {
+  if (!empty($_GET['debug'])) {
+    drupal_set_message($msg);
+  }
+}
+
+/**
+ * Get local elements from a local list.
+ */
```
Comment (cableman) [inline modules/p2/ding_list/modules/ding_listrescue/ding_listrescue.module]: Missing @param, @return function documentation.

File: modules/p2/ding_list/modules/ding_listrescue/ding_listrescue.module [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+  // Weight is used by ratings.
+  $element_query->addField('fw', 'field_weight_value', 'weight');
+
+  return $element_query->range(0, $max_elements)->execute()->fetchAll();
+}
+
+/**
+ * Get all the local lists.
+ */
+function ding_listrescue_get_local_lists($account = NULL, $types = NULL) {
```
Comment (cableman) [inline modules/p2/ding_list/modules/ding_listrescue/ding_listrescue.module]: Missing @param, @return function documentation.

File: modules/p2/ding_list/modules/ding_listrescue/ding_listrescue.module [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+    ->condition('dt.uid', $account->uid)
+    ->condition('flt.field_list_type_value', $types);
+  $list_query->addField('flt', 'field_list_type_value', 'type');
+
+  return $list_query->range(0, $max_lists)->execute()->fetchAll();
+}
+
+/**
+ * Get a matching remote list.
+ */
```
Comment (cableman) [inline modules/p2/ding_list/modules/ding_listrescue/ding_listrescue.module]: Missing @param, @return function documentation.

**Comment by GitCop** [comment]
Please ensure that your commit messages comply with our contribution guidelines.

 * Commit: 57497e51762f90fa0f8c78632b37c806a2e97208
  * The first line of your commit message (the subject) must follow this format: [issue number]: [short summary of change]

Guidelines are available at https://github.com/ding2/ding2/blob/master/CONTRIBUTING.md.

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
Please ensure that your commit messages comply with our contribution guidelines.

 * Commit: 57497e51762f90fa0f8c78632b37c806a2e97208
  * The first line of your commit message (the subject) must follow this format: [issue number]: [short summary of change]

Guidelines are available at https://github.com/ding2/ding2/blob/master/CONTRIBUTING.md.

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
Please ensure that your commit messages comply with our contribution guidelines.

 * Commit: 4a9ee51fb512fea63d5081646421c152d7fa313a
  * The first line of your commit message (the subject) must follow this format: [issue number]: [short summary of change]

Guidelines are available at https://github.com/ding2/ding2/blob/master/CONTRIBUTING.md.

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
Please ensure that your commit messages comply with our contribution guidelines.

 * Commit: 4a9ee51fb512fea63d5081646421c152d7fa313a
  * The first line of your commit message (the subject) must follow this format: [issue number]: [short summary of change]

Guidelines are available at https://github.com/ding2/ding2/blob/master/CONTRIBUTING.md.

---
This message was auto-generated by https://gitcop.com

**Comment by martinb14** [comment]
Fixed @cableman change requests

File: modules/p2/ding_list/modules/ding_listrescue/ding_listrescue.module (hunk 3 above)
Comment (cableman) [reply modules/p2/ding_list/modules/ding_listrescue/ding_listrescue.module]: Please see the first question? Why this debug statement is here?

**Comment by cableman** [comment]
Please answer the question to ensure at each comment to ensure that you actually have seen them all and to speed up the review process.

Summery (questions now answered):
1. Why the debug logger message function?
2. Why the extra user_loads, when you already have the global user object?
3. Stil why the limit (setting them higher do not answer the question)?

**Comment by martinb14** [comment]
1) Logger removed
2) user_load removed (Previous bug in Drupal required this in some cases)
3) Comments describes why the limit.

**Pull request merged**